### ISSUE TITLE
Better Manual Broadcasting for ufuncs

### ISSUE THREAD

**Issue by randolf-scholz** [numpy/numpy#12900]
Standard ufuncs like add, multiply and so on should allow for manual broadcasting. In particular, I suggest adding support of the [ufunc axes keyword](https://docs.scipy.org/doc/numpy-1.15.1/reference/ufuncs.html#ufuncs-kwargs) to add the math functions listed [here.](https://docs.scipy.org/doc/numpy-1.15.1/reference/ufuncs.html#math-operations) (My apologies if this is already in the works, as far as I can tell it doesn't work in 1.15.4)
 
The main issue with the status quo is that it makes it unnecessarily complicated and awkward to write routines that a priori do not know the exact shapes of incoming tensors. (but may know it partially, for example it could be known that the incoming tensor has the shape (?,3,3), where the ? could contain any number of additional axes. )

### Reproducing code example:

For example a task as simple as "Add vector v to the tensor T along the axis k" is really awkward to implement:

```python
import numpy as np
T = np.ones((2,2,3,3))
v = np.arange(3)
# z = T + v                                     # adds along the last axis, not what we want!
# z = T + v[None, None, : , None] # does work, but incredibly ugly / hard to read
# z = np.add(T, v, axes=([2],[0]))   # suggested behaviour
```

**Comment by mhvk** [comment]
@randolf-scholz - thanks for your suggestion! I fear this is quite counter, though, to the whole idea that ufunc operate on single elements, i.e., their implementation works on individual elements, and that broadcasting works exactly as it does when you just write `T + v`. With your suggestion, we would be letting what is supposed to be a very simple function also implement `reshape` and `transpose` for its arguments. I also find it troublesome that it assumes full knowledge of all the axes - in much code, it is very useful if things also work if someone puts in an item with more axes; numpy broadcasting allows you to do that as long as those axes are at the front. (Note that this is different for reductions and generalized ufuncs, where operation is not over single elements but rather over one or more dimensions, and where one can indeed indicate which axes are supposed to be used).

Note that the solution to at least the first example is a bit less ugly/hard to read than you suggest:
```
z = T + v[:, np.newaxis]  # if you *know* v can only have one axis
z = T + v[..., np.newaxis]  # safer version that allows v to have more than one axis
```
Especially in the latter case I am only assuming that the last axis of `T` is special, in that `v` should always be broadcast against it.

For your second example, it might be one where the complexity suggests that axes are not ordered logically - if you want stacks of 3x3 matrices, it would make more sense to have those as the last elements, since this is the assumption in matrix multiplication `@` (and our code has been optimized for that memory layout).

**Comment by randolf-scholz** [comment]
@mhvk Thanks for your reply. 

>  I also find it troublesome that it assumes full knowledge of all the axes - in much code, it is very useful if things also work if someone puts in an item with more axes

Well, this is actually one of the reasons for my suggestion. And I don't think that actually full knowledge of the axes is needed. When writing `np.add(T, v, axes=([2],[0]))` it should really just translate to "add v_i to each slice of T along its 3rd axis", so it it agnostic about how many axis T has (other than that there are >=3)

>  numpy broadcasting allows you to do that as long as those axes are at the front.

Yes but this is actually one of my main grievances with numpy. Often When working with data tensors one wants to perform operations along specific axes. Since numpy doesn't allow specifying the axis for many elementary operations, one end up either doing:

A: Reshaping / transposing, such that the target axis one wants to operate on is the last one and broadcasting works

B: Using the, in my opinion hideous, `np.newaxis` method

> Note that the solution to at least the first example is a bit less ugly/hard to read than you suggest: `T + v[..., np.newaxis]`

While this method works in this example, consider the following: Let T be a Tensor of unknown rank >=k, and v be a vector compatible with the k-th axis of T. Then afaik the only real ways to implement this are the following:

```python
z = np.swapaxes(T, k, -1) + v
z = np.swapaxes(z, k, -1)
```
or the incredible

```python
z = T + v[tuple(None if i != 2 else slice(None) for i in range(T.ndim))]
```
Of course, in more complex examples involving working over several axes this does not work either.  And again, think about how much cleaner, concise, and readable a simple

```python
z = np.add(T, v, axes=(k,0))
```
would be.

> For your second example, it might be one where the complexity suggests that axes are not ordered logically - if you want stacks of 3x3 matrices, it would make more sense to have those as the last elements, since this is the assumption in matrix multiplication @ (and our code has been optimized for that memory layout).

The examples were deliberately arbitrary. The point is that a linear-algebra toolbox whose main datatype is a tensor should allow me to easily do any computation I could write down in Einstein notation.

**Comment by mhvk** [comment]
@randolf-scholz - I see that your suggestion is handy for some situations, but I still think it is makes things more complicated for the general case. E.g., what to do when both operands have more axes than the ones you are enumerating - where do the "remaining" axes go? And what is the default output axis ordering?

Anyway, this feature request is big enough that I would suggest to propose it on the numpy-dev mailing list (where all more significant API additions/changes are discussed). I have worked enough on ufunc internals to know it would require a fair amount of work, so best to get a sense of whether others feel this is a good idea before time is spent by you or anyone else trying to implement it.

**Comment by seberg** [comment]
I agree with @mhvk. I see that it does try to solve a real usability issue. However, currently I am not sure that the current suggestion is very elegant/obvious.

IIRC, we previously had possibly similar suggestions of having something like an `np.einops` similar to `np.einsum` to allow different ufuncs to replace the multiplication and addition of normal einsum.

There is also some push to add multiple axes to `np.expand_dims` which may make this a bit less tedious to write (note that this is also a two line helper function you could define yourself), such as:
```
def align_vector(vector, dim):
    return vector[mildly_annoying_indexing_operation]   # or reshape
```

**Comment by eric-wieser** [comment]
I suspect this is a use-case where @shoyer's xarray really shines.

**Comment by shoyer** [comment]
Xarray does indeed work well for cases like this -- you can give names to axes on arrays (dimensions) which are used for broadcasting instead of position.

That said, I like the look of the `broadcasting` keyword argument for controlling this explicitly with `einsum` like notation in NumPy -- this is certainly far better than awkward uses of `reshape` and `newaxis` (I think `axes` would be confusing).

I do like the idea of writing a separate utility function inspired by `einsum()` first. Something like:
```
x, y = np.custom_broadcast('ijkl,k->ijkl', x, y)
```
Then a ufunc `broadcasting` keyword could reuse this same logic. I think this is roughly compatibility with the ufunc/broadcasting model, since all these axes reorderings and insertions can be done by only manipulating strides/shape without copying any data.

There would be quite a few design decisions to make for specifying exactly how this works, e.g.,
- should we allow for broadcasting different arguments to different shapes? If so, we might require specifying multiple outputs in the string, e.g., something like `'ijkl,k->ijkl,ijkl'`.
- would a new index on the output be an error or mean inserting a dimension of size 1? e.g., `ab->abc`

**Comment by eric-wieser** [comment]
Another consideration would be to extend `ndarray.transpose` to accept `np.newaxis`. Comparing:

Now | Proposed
------|---------
`a_3d.transpose((2, 1, 0))[:, np.newaxis, :, :]` | `a_4d.tranpose((2, np.newaxis, 1, 0))`
`a_4d[:, np.newaxis, :, :]` | `a_4d.tranpose((0, np.newaxis, 1, 2))` - not any clearer, but at least more consistent

**Comment by mhvk** [comment]
@eric-wieser - that looks useful. 

At the risk of "more than one method", `transpose` might also gain the ability to interpret strings where one could use `1`, as in `a_2d.tranpose('k1ji')` ... 

hmmm...  seeing that, actually not sure about that one.  

But I do like the `np.newaxis` extension - it seems natural. Worth its own issue?
